File: src/lib/tools/image-stego/outguessWasm.ts
/**
 * OutGuess payload extractor backed by a real Emscripten build of the reference OutGuess 0.2 C
 * source (see tools/outguess-wasm/), rather than a reimplementation — byte-for-byte verified against
 * the canonical Cicada 3301 sample (boxentriq.com/samples/outguess-sample.jpg) and boxentriq's own
 * emcc-compiled outguess.wasm output.
 *
 * outguess.c's CLI declares `char *key = "Default key"` and only overwrites it if the user passes
 * `-k <key>` — an empty/omitted passphrase is NOT the same as an empty key. This module reproduces
 * that: a blank passphrase is sent to the wasm module as the literal bytes "Default key", matching
 * every real-world file embedded without an explicit passphrase (the common case).
 *
 * A fresh module instance is instantiated per extraction call rather than reused: outguess's C code
 * calls exit(1) directly on malformed input (wrong passphrase, no payload, corrupt header) — this is
 * unmodified upstream behavior in steg_retrieve()/decode_data(), not a bug introduced here. Emscripten
 * turns that into a catchable thrown JS exception, but the module's internal state afterward isn't
 * guaranteed reusable, so each call gets its own isolated instance instead of trying to recover one.
 */
import type { OutguessExtractResult } from "./outguess";

const OUTGUESS_DEFAULT_KEY = "Default key";

interface OutguessWasmExports {
  FS: { writeFile: (path: string, data: Uint8Array) => void };
  HEAPU8: Uint8Array;
  HEAP32: Int32Array;
  _malloc: (size: number) => number;
  _free: (ptr: number) => void;
  _wasm_outguess_free: (ptr: number) => void;
  stringToNewUTF8: (s: string) => number;
  ccall: (name: string, returnType: string, argTypes: string[], args: unknown[]) => number;
}

type OutguessModuleFactory = (opts?: Record<string, unknown>) => Promise<OutguessWasmExports>;

let moduleFactory: OutguessModuleFactory | null | undefined;

/**
 * Lazily loads the wasm module factory. Returns null (once) if the asset isn't available.
 *
 * The glue script lives in public/wasm/, so it's served as-is and must NOT be reached via a plain
 * `import("/wasm/outguess.js")` — Vite's dev server refuses to run its module transform pipeline on
 * anything under /public ("should not be imported from source code"). Fetching the source as text
 * and importing it via a blob: URL sidesteps that, in both dev and prod.
 */
async function loadModuleFactory(): Promise<OutguessModuleFactory | null> {
  if (moduleFactory !== undefined) return moduleFactory;
  try {
    const source = await fetch("/wasm/outguess.js").then((r) => {
      if (!r.ok) throw new Error(`outguess.js fetch failed: ${r.status}`);
      return r.text();
    });
    const blobUrl = URL.createObjectURL(new Blob([source], { type: "text/javascript" }));
    try {
      const mod = await import(/* @vite-ignore */ blobUrl);
      moduleFactory = mod.default as OutguessModuleFactory;
    } finally {
      URL.revokeObjectURL(blobUrl);
    }
  } catch {
    moduleFactory = null;
  }
  return moduleFactory;
}

/** Options passed to every module instantiation, so the wasm binary resolves relative to /wasm/
 * rather than to the blob: URL the glue script was loaded from. */
function instantiateOpts() {
  return { locateFile: (path: string) => `/wasm/${path}` };
}

/** Whether the compiled OutGuess wasm module is available in this build (vs. only the JS fallback). */
export async function isOutguessWasmAvailable(): Promise<boolean> {
  return (await loadModuleFactory()) !== null;
}

/**
 * Attempts to extract an OutGuess payload from `file` using `passphrase` (may be empty — mapped to
 * outguess's own "Default key" default). Returns null if the wasm module itself isn't available
 * (asset missing / failed to load), so the caller can fall back to the pure-JS heuristic extractor.
 */
export async function extractOutguessWasm(file: File, passphrase: string): Promise<OutguessExtractResult | null> {
  const factory = await loadModuleFactory();
  if (!factory) return null;

  const effectiveKey = passphrase.length > 0 ? passphrase : OUTGUESS_DEFAULT_KEY;
  const keyBytes = new TextEncoder().encode(effectiveKey);
  const fileBytes = new Uint8Array(await file.arrayBuffer());

  let mod: OutguessWasmExports;
  try {
    mod = await factory(instantiateOpts());
  } catch {
    return null;
  }

  mod.FS.writeFile("/input.jpg", fileBytes);

  const keyPtr = mod._malloc(Math.max(1, keyBytes.length));
  mod.HEAPU8.set(keyBytes, keyPtr);
  const outLenPtr = mod._malloc(4);
  const pathPtr = mod.stringToNewUTF8("/input.jpg");

  let resultPtr = 0;
  try {
    resultPtr = mod.ccall(
      "wasm_outguess_extract",
      "number",
      ["number", "number", "number", "number"],
      [pathPtr, keyPtr, keyBytes.length, outLenPtr]
    );
  } catch {
    // outguess's own code calls exit(1) on a wrong passphrase / no embedded payload / corrupt
    // header — this is the expected shape of a "nothing found" result, not an unusual failure.
    return {
      success: false,
      error: "No OutGuess payload found — wrong passphrase, or this JPEG has no embedded data."
    };
  }

  if (resultPtr === 0) {
    return { success: false, error: "Unsupported cover format for OutGuess — needs a JPEG that libjpeg can decode." };
  }

  const outLen = mod.HEAP32[outLenPtr / 4];
  const bytes = mod.HEAPU8.slice(resultPtr, resultPtr + outLen);
  mod._wasm_outguess_free(resultPtr);

  let text: string | undefined;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    text = undefined;
  }

  return { success: true, bytes, text };
}

File: tools/outguess-wasm/wrapper/outguess_wasm.c
/*
 * Emscripten wrapper exposing OutGuess 0.2's extraction path as a small C API, so the JS/TS side
 * can call a single function instead of driving outguess's CLI (main() in outguess.c parses argv
 * and calls exit() on error, which we don't want in a library build).
 *
 * Build approach (see ../build.sh):
 *   - Compiled together with the unmodified upstream sources in ../vendor/outguess-src (arc.c,
 *     iterator.c, outguess.c, jpg.c, golay.c, md5.c, err.c) and the plain libjpeg 6b library object
 *     files from ../vendor/outguess-src/jpeg-6b-steg (NOT the cjpeg/djpeg/jpegtran CLI files).
 *   - outguess.c's own main() is left in the link (harmless — it's simply never called; we don't
 *     pass -sINVOKE_RUN, and the file's `main` symbol isn't in EXPORTED_FUNCTIONS).
 *   - Input file bytes are written to Emscripten's MEMFS by the JS side (FS.writeFile) before
 *     calling wasm_outguess_extract(), so jpg.c's read_JPEG_file(FILE*) works completely unchanged
 *     via a normal fopen() — no need to patch libjpeg's file I/O for in-memory buffers.
 *
 * Verified against the canonical Cicada 3301 sample (boxentriq.com/samples/outguess-sample.jpg,
 * empty passphrase): decodes to the expected 535-byte "Here is a book code..." plaintext,
 * bit-for-bit matching boxentriq's own emcc-compiled outguess.wasm.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

#include "outguess.h"
#include "arc.h"
#include "iterator.h"
#include "pnm.h"
#include "jpg.h"

/* jpg.c's format handler (declared `extern handler jpg_handler;` in jpg.h). */

/*
 * Extracts an OutGuess payload from the JPEG at `inputPath` (write it there via FS.writeFile first)
 * using `key`/`keylen`. Callers must pass outguess's own CLI default key, the literal bytes
 * "Default key" (11 bytes, no NUL), when the user hasn't supplied a passphrase — outguess.c's
 * main() declares `char *key = "Default key"` and only overwrites it if `-k` is given, so an
 * empty/no passphrase is NOT the same as an empty key. Getting this wrong silently produces
 * "implausible header" failures even on valid files (this cost real debugging time — verified via
 * a working extraction only once "Default key" was tried in place of "").
 *
 * On success, *outLen is set and the return value is a malloc'd buffer the caller must free with
 * wasm_outguess_free(). Returns NULL on unreadable-JPEG failure.
 *
 * IMPORTANT — steg_retrieve()/decode_data() inside outguess.c call exit(1) directly on malformed
 * input (wrong passphrase, no embedded payload, corrupt header) rather than returning an error —
 * this is upstream, unmodified behavior, not something introduced here. Emscripten turns exit()
 * into a thrown JS `ExitStatus` exception that unwinds out of the ccall boundary (confirmed safe to
 * catch in JS), but the module instance's internal state after that point is not guaranteed usable
 * for a second call. Callers MUST instantiate a fresh module per extraction attempt rather than
 * reusing one across calls — see outguessWasm.ts.
 */
unsigned char *wasm_outguess_extract(const char *inputPath, const unsigned char *key, int keylen, int *outLen) {
	FILE *fin = fopen(inputPath, "rb");
	if (!fin) {
		*outLen = 0;
		return NULL;
	}

	image *img = jpg_handler.read(fin);
	fclose(fin);
	if (!img) {
		*outLen = 0;
		return NULL;
	}

	bitmap bmap;
	jpg_handler.get_bitmap(&bmap, img, STEG_RETRIEVE);

	struct arc4_stream as, tas;
	iterator iter;
	int len = 0;

	arc4_initkey(&as, "Encryption", (u_char *)key, keylen);
	/*
	 * outguess.c's main() takes this snapshot right after arc4_initkey, before steg_retrieve()
	 * consumes keystream bytes from `as` to decode the 4-byte header — decode_data() below must
	 * start from the pre-header state (`tas`), not the state `as` is left in after steg_retrieve.
	 */
	tas = as;
	iterator_init(&iter, &bmap, (u_char *)key, keylen);

	char *encdata = steg_retrieve(&len, &bmap, &iter, &as, 0 /* flags: no ECC by default */);
	if (!encdata) {
		*outLen = 0;
		return NULL;
	}

	int declen = len;
	unsigned char *data = decode_data((u_char *)encdata, &declen, &tas, 0);
	free(encdata);

	*outLen = declen;
	return data;
}

void wasm_outguess_free(void *ptr) {
	free(ptr);
}
